embedded-graphics/src/draw_target.rs

```diff
@@ -0,0 +1,304 @@
+    primitives::{self, Primitive},
+    style::{PrimitiveStyle, Styled},
+};
+
+/// Defines a display that can be used to render [`Drawable`] objects.
+///
+/// To use this crate in a driver, `DrawTarget` must be implemented. This trait defines how a
+/// display draws pixels, and optionally provides a way to define accelerated drawing methods for
+/// graphical primitives such as lines, rectangles, triangles, and circles.
+///
+/// Once a `DrawTarget` is defined, it can be used to render [`Drawable`]s. Note that any iterator
+/// over [`Pixel`]s has a default implementation for the [`Drawable`] trait. See the [`Drawable`]
+/// trait documentation for more details.
+///
+/// Here's an example for an imaginary display that has a 64x64px framebuffer of 8 bit values that
+/// communicates over a (simplified) SPI interface:
+///
+/// ```rust
+/// use embedded_graphics::{
+///     drawable::Pixel,
+///     egcircle,
+///     geometry::Size,
+///     pixelcolor::{Gray8, GrayColor},
+///     prelude::*,
+///     primitive_style, DrawTarget,
+/// };
+///
+/// # struct SPI1;
+/// #
+/// # impl SPI1 {
+/// #     pub fn send_bytes(&self, buf: &[u8]) -> Result<(), ()> {
+/// #         Ok(())
+/// #     }
+/// # }
+/// #
+/// /// A fake display 64px x 64px where each pixel is stored as a single `u8`
+/// struct ExampleDisplay {
+///     framebuffer: [u8; 64 * 64],
+///     iface: SPI1,
+/// }
+///
+/// impl ExampleDisplay {
+///     /// Send buffer to the display
+///     pub fn flush(&self) -> Result<(), ()> {
+///         self.iface.send_bytes(&self.framebuffer)
+///     }
+/// }
+///
+/// impl DrawTarget<Gray8> for ExampleDisplay {
+///     type Error = core::convert::Infallible;
+///
+///     /// Draw a `Pixel` that has a color defined as `Gray8`.
+///     fn draw_pixel(&mut self, pixel: Pixel<Gray8>) -> Result<(), Self::Error> {
+///         let Pixel(coord, color) = pixel;
+///         // Place an (x, y) pixel at the right index in the framebuffer
+///         let index = coord.x + coord.y * 64;
+///         self.framebuffer[index as usize] = color.luma();
+///
+///         Ok(())
+///     }
+///
+///     fn size(&self) -> Size {
+///         Size::new(64, 64)
+///     }
+/// }
+///
+/// let mut display = ExampleDisplay {
+///     framebuffer: [0; 4096],
+///     iface: SPI1,
+/// };
+///
+/// // Draw a circle centered around `(32, 32)` with a radius of `10` and a white stroke
+/// let circle = egcircle!(
+///     center = (32, 32),
+///     radius = 10,
+///     style = primitive_style!(stroke_color = Gray8::WHITE)
+/// );
+/// circle.draw(&mut display)?;
+///
+/// // Update the display
+/// display.flush().expect("Failed to send data to display");
+/// # Ok::<(), core::convert::Infallible>(())
+/// ```
+///
+/// ## Hardware Acceleration
+///
+/// In addition to defining [`draw_pixel`], an implementation of [`DrawTarget`] can also provide
+/// alternative implementations of drawing methods for graphical primitives. Here is an example of
```

Comment:
Should we use "styled primitives" or "styled graphical primitives" here, because primitives aren't drawable on their own?